pkg/storage/etcd3/watch_test.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type testWatchStruct struct {
+	obj         *api.Pod
+	expectEvent bool
+}
+
+// TestWatchAdd tests Watch of first occurrence of objects should notify Add event.
+func TestWatchAdd(t *testing.T) {
+	ctx, store, cluster := testSetup(t)
+	defer cluster.Terminate(t)
+
+	podFoo := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}}
+	podBar := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "bar"}}
+
+	tests := []struct {
```

Comment:
Table driven tests?  It's not an anti-pattern in Go, as long as the tests
are clearly exercising inputs and outputs of a standard function.

On Wed, Apr 13, 2016 at 11:34 AM, Timothy St. Clair <
notifications@github.com> wrote:

> In pkg/storage/etcd3/watch_test.go
> <https://github.com/kubernetes/kubernetes/pull/23694#discussion_r59570597>
> :
>
> > +)
> > +
> > +type testWatchStruct struct {
> > +	obj         *api.Pod
> > +	expectEvent bool
> > +}
> > +
> > +// TestWatchAdd tests Watch of first occurrence of objects should notify Add event.
> > +func TestWatchAdd(t *testing.T) {
> > +	ctx, store, cluster := testSetup(t)
> > +	defer cluster.Terminate(t)
> > +
> > +	podFoo := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}}
> > +	podBar := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "bar"}}
> > +
> > +	tests := []struct {
>
> I know this is a standard pattern in the tests, but I've always found it
> obtuse and almost an anti-pattern for folks trying to learn the behavior of
> the system.
>
> I try to rid this pattern every time I dive into something.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23694/files/5385e73b99f3754dc904eadf172308d62169a4da#r59570597>
>